subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileCollectionSnapshot.java

```diff
@@ -41,6 +41,11 @@
      */
     Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, String title, boolean includeAdded);
 
+    /**
+     * Returns the combined hash of the contents.
+     *
+     * We short-circuit {@link #iterateContentChangesSince(FileCollectionSnapshot, String, boolean)} if we already know the hash.
```

Comment:
We only short circuit if we know the hash _and_ it's equal to the hash of the file collection we compare it to.